File: android/engine/src/main/kotlin/ru/gigapisar/engine/Rnnt.kt
package ru.gigapisar.engine

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import java.io.File
import java.nio.FloatBuffer
import java.nio.LongBuffer

/**
 * Распознавание речи: звук → признаки → энкодер → жадное декодирование RNN-T → текст.
 * Повторяет server/giga_core.py и web/giga/recognizer.js шаг в шаг; тот же
 * пакет ai.onnxruntime работает и на JVM (сверка), и на Android.
 */
class Rnnt private constructor(
    private val env: OrtEnvironment,
    val config: ModelConfig,
    private val encoder: OrtSession,
    private val decoder: OrtSession,
    private val joint: OrtSession,
    val tokenizer: Tokenizer,
) : AutoCloseable {
    class ModelConfig(val features: FeatureConfig, val predHidden: Int, val predLayers: Int)

    private val features = Features(config.features)
    val sampleRate get() = config.features.sampleRate
    private val encIn = encoder.inputInfo.keys.toList()
    private val encOut = encoder.outputInfo.keys.toList()
    private val decIn = decoder.inputInfo.keys.toList()
    private val decOut = decoder.outputInfo.keys.toList()
    private val jIn = joint.inputInfo.keys.toList()

    /** Распознаёт запись любой длины: длинная режется по паузам и склеивается. */
    fun transcribe(samples: FloatArray): String {
        val rate = sampleRate
        val total = samples.size.toDouble() / rate
        if (total <= MAX_CHUNK + 1) return transcribeWave(samples)
        val parts = ArrayList<String>()
        for ((a, b) in Chunker.chunkBounds(total, Chunker.silences(samples, rate), MAX_CHUNK)) {
            val from = minOf(samples.size, (a * rate).toInt()); val to = minOf(samples.size, (b * rate).toInt())
            if (to > from) transcribeWave(samples.copyOfRange(from, to)).takeIf { it.isNotEmpty() }?.let(parts::add)
        }
        return parts.joinToString(" ")
    }

    /** Одна волна не длиннее предела модели. */
    @Synchronized
    fun transcribeWave(wave: FloatArray): String {
        val f = features.compute(wave)
        if (f.frames <= 0) return ""
        val nMels = config.features.nMels.toLong()
        OnnxTensor.createTensor(env, FloatBuffer.wrap(f.values), longArrayOf(1, nMels, f.frames.toLong())).use { x ->
            OnnxTensor.createTensor(env, LongBuffer.wrap(longArrayOf(features.outLen(wave.size).toLong())), longArrayOf(1)).use { len ->
                encoder.run(mapOf(encIn[0] to x, encIn[1] to len)).use { out ->
                    val enc = out[0] as OnnxTensor
                    val shape = enc.info.shape
                    val encD = shape[1].toInt(); val encT = shape[2].toInt()
                    val encLen = minOf(lengthOf(out[1] as OnnxTensor) ?: encT, encT)
                    val ids = greedy(enc.floatBuffer, encD, encT, encLen)
                    return tokenizer.decode(ids)
                }
            }
        }
    }

    /**
     * Жадное декодирование. Выход декодера зависит только от последней буквы
     * и состояния, поэтому считается заново лишь когда буква выдана.
     */
    private fun greedy(encoded: FloatBuffer, encD: Int, encT: Int, encLen: Int): IntArray {
        val blank = tokenizer.blankId
        val layers = config.predLayers; val hidden = config.predHidden
        val stateShape = longArrayOf(layers.toLong(), 1, hidden.toLong())
        val zeros = FloatArray(layers * hidden)
        val hyp = ArrayList<Int>()
        var label = blank
        var h = zeros; var c = zeros
        var started = false
        var g: FloatArray? = null; var hNext = zeros; var cNext = zeros
        val frame = FloatArray(encD)

        fun runDecoder() {
            val lab = OnnxTensor.createTensor(env, LongBuffer.wrap(longArrayOf((if (started) label else blank).toLong())), longArrayOf(1, 1))
            val hT = OnnxTensor.createTensor(env, FloatBuffer.wrap(if (started) h else zeros), stateShape)
            val cT = OnnxTensor.createTensor(env, FloatBuffer.wrap(if (started) c else zeros), stateShape)
            try {
                decoder.run(mapOf(decIn[0] to lab, decIn[1] to hT, decIn[2] to cT)).use { r ->
                    g = floats(r[0] as OnnxTensor); hNext = floats(r[1] as OnnxTensor); cNext = floats(r[2] as OnnxTensor)
                }
            } finally { lab.close(); hT.close(); cT.close() }
        }

        for (t in 0 until encLen) {
            for (d in 0 until encD) frame[d] = encoded.get(d * encT + t)
            OnnxTensor.createTensor(env, FloatBuffer.wrap(frame), longArrayOf(1, encD.toLong(), 1)).use { frameT ->
                for (s in 0 until MAX_SYMBOLS_PER_FRAME) {
                    if (g == null) runDecoder()
                    val gT = OnnxTensor.createTensor(env, FloatBuffer.wrap(g!!), longArrayOf(1, hidden.toLong(), 1))
                    val best: Int
                    try {
                        joint.run(mapOf(jIn[0] to frameT, jIn[1] to gT)).use { r ->
                            val logits = (r[0] as OnnxTensor).floatBuffer
                            var bi = 0; var bv = Float.NEGATIVE_INFINITY
                            val n = logits.remaining()
                            for (i in 0 until n) { val v = logits.get(i); if (v > bv) { bv = v; bi = i } }
                            best = bi
                        }
                    } finally { gT.close() }
                    if (best == blank) break
                    hyp.add(best)
                    label = best; h = hNext; c = cNext; started = true
                    g = null
                }
            }
        }
        return hyp.toIntArray()
    }

    override fun close() { encoder.close(); decoder.close(); joint.close() }

    companion object {
        const val MODEL_NAME = "v3_e2e_rnnt"
        val FILES = listOf("$MODEL_NAME.yaml", "${MODEL_NAME}_encoder.onnx", "${MODEL_NAME}_decoder.onnx",
            "${MODEL_NAME}_joint.onnx", "${MODEL_NAME}_tokenizer.model")
        const val MAX_CHUNK = 24.0            // предел одного прохода модели — 25 секунд
        const val MAX_SYMBOLS_PER_FRAME = 3

        /** Длина выхода энкодера: int64 или int32 — смотря как экспортирована модель. */
        private fun lengthOf(t: OnnxTensor): Int? = when (t.info.type) {
            ai.onnxruntime.OnnxJavaType.INT64 -> t.longBuffer?.takeIf { it.remaining() > 0 }?.get(0)?.toInt()
            ai.onnxruntime.OnnxJavaType.INT32 -> t.intBuffer?.takeIf { it.remaining() > 0 }?.get(0)
            else -> null
        }

        private fun floats(t: OnnxTensor): FloatArray { val b = t.floatBuffer; val a = FloatArray(b.remaining()); b.get(a); return a }

        fun hasModel(dir: File) = FILES.all { File(dir, it).length() > 0 }

        /** Разбор нужных строк yaml — все ключи уникальны по смыслу. */
        fun parseConfig(text: String): ModelConfig {
            fun value(key: String) = text.lineSequence().map { it.trim() }.firstOrNull { it.startsWith("$key:") }?.substring(key.length + 1)?.trim()
            fun int(key: String, def: Int) = value(key)?.toIntOrNull() ?: def
            val d = FeatureConfig()
            return ModelConfig(
                FeatureConfig(int("sample_rate", d.sampleRate), int("features", d.nMels), int("n_fft", d.nFFT),
                    int("win_length", d.winLength), int("hop_length", d.hopLength), value("center")?.let { it == "true" } ?: d.center),
                int("pred_hidden", 320), int("pred_rnn_layers", 1),
            )
        }

        /** Загружает модель из папки с пятью файлами. [threads] — потоков на энкодер. */
        fun load(dir: File, threads: Int = 4): Rnnt {
            val env = OrtEnvironment.getEnvironment()
            fun opts() = OrtSession.SessionOptions().apply {
                setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
                setIntraOpNumThreads(threads)
            }
            val cfg = parseConfig(File(dir, FILES[0]).readText())
            val encoder = env.createSession(File(dir, FILES[1]).path, opts())
            // декодер и джойнт крошечные, их гоняют тысячи раз: один поток быстрее, чем раздача работы
            val small = OrtSession.SessionOptions().apply { setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT); setIntraOpNumThreads(1) }
            val decoder = env.createSession(File(dir, FILES[2]).path, small)
            val joint = env.createSession(File(dir, FILES[3]).path, small)
            return Rnnt(env, cfg, encoder, decoder, joint, Tokenizer(File(dir, FILES[4]).readBytes()))
        }
    }
}

File: windows/gigapisar/asr/rnnt.go
package asr

import (
	"errors"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"

	"gigapisar/ort"
)

const (
	ModelName          = "v3_e2e_rnnt"
	MaxChunk           = 24.0 // предел одного прохода модели — 25 секунд
	maxSymbolsPerFrame = 3
)

// Files — пять файлов модели GigaAM v3.
var Files = []string{ModelName + ".yaml", ModelName + "_encoder.onnx", ModelName + "_decoder.onnx", ModelName + "_joint.onnx", ModelName + "_tokenizer.model"}

func HasModel(dir string) bool {
	for _, f := range Files {
		if st, err := os.Stat(filepath.Join(dir, f)); err != nil || st.Size() == 0 {
			return false
		}
	}
	return true
}

type ModelConfig struct {
	Features               FeatureConfig
	PredHidden, PredLayers int
}

func ParseConfig(text string) ModelConfig {
	value := func(key string) (string, bool) {
		for _, line := range strings.Split(text, "\n") {
			s := strings.TrimSpace(line)
			if strings.HasPrefix(s, key+":") {
				return strings.TrimSpace(s[len(key)+1:]), true
			}
		}
		return "", false
	}
	num := func(key string, def int) int {
		if v, ok := value(key); ok {
			if n, err := strconv.Atoi(v); err == nil {
				return n
			}
		}
		return def
	}
	d := DefaultFeatures
	c, hasC := value("center")
	return ModelConfig{
		Features: FeatureConfig{SampleRate: num("sample_rate", d.SampleRate), NMels: num("features", d.NMels), NFFT: num("n_fft", d.NFFT),
			WinLength: num("win_length", d.WinLength), HopLength: num("hop_length", d.HopLength), Center: hasC && c == "true"},
		PredHidden: num("pred_hidden", 320), PredLayers: num("pred_rnn_layers", 1),
	}
}

// Recognizer — энкодер, декодер, джойнт и токенизатор; жадное декодирование RNN-T.
type Recognizer struct {
	Config    ModelConfig
	enc, dec  *ort.Session
	joint     *ort.Session
	tok       *Tokenizer
	feat      *Features
	mu        sync.Mutex
}

// Load грузит модель из папки (onnxruntime должен быть загружен: ort.Load).
func Load(dir string, threads int) (*Recognizer, error) {
	read := func(name string) ([]byte, error) { return os.ReadFile(filepath.Join(dir, name)) }
	yaml, err := read(Files[0])
	if err != nil {
		return nil, err
	}
	cfg := ParseConfig(string(yaml))
	r := &Recognizer{Config: cfg, feat: NewFeatures(cfg.Features)}
	for i, dst := range []**ort.Session{&r.enc, &r.dec, &r.joint} {
		b, err := read(Files[1+i])
		if err != nil {
			return nil, err
		}
		th := 1 // декодер и джойнт крошечные: один поток быстрее раздачи работы
		if i == 0 {
			th = threads
		}
		s, err := ort.NewSession(b, th)
		if err != nil {
			return nil, fmt.Errorf("%s: %w", Files[1+i], err)
		}
		*dst = s
	}
	tb, err := read(Files[4])
	if err != nil {
		return nil, err
	}
	if r.tok, err = NewTokenizer(tb); err != nil {
		return nil, err
	}
	return r, nil
}

func (r *Recognizer) Close() {
	for _, s := range []*ort.Session{r.enc, r.dec, r.joint} {
		if s != nil {
			s.Close()
		}
	}
}

func (r *Recognizer) SampleRate() int { return r.Config.Features.SampleRate }

// Transcribe — запись любой длины: длинная режется по паузам и склеивается.
func (r *Recognizer) Transcribe(samples []float32) (string, error) {
	rate := r.SampleRate()
	total := float64(len(samples)) / float64(rate)
	if total <= MaxChunk+1 {
		return r.TranscribeWave(samples)
	}
	var parts []string
	for _, b := range ChunkBounds(total, Silences(samples, rate, -35, 0.3), MaxChunk) {
		from, to := int(b[0]*float64(rate)), int(b[1]*float64(rate))
		if to > len(samples) {
			to = len(samples)
		}
		if to <= from {
			continue
		}
		t, err := r.TranscribeWave(samples[from:to])
		if err != nil {
			return "", err
		}
		if t != "" {
			parts = append(parts, t)
		}
	}
	return strings.Join(parts, " "), nil
}

// TranscribeWave — одна волна не длиннее предела модели.
func (r *Recognizer) TranscribeWave(wave []float32) (string, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	values, frames := r.feat.Compute(wave)
	if frames <= 0 {
		return "", nil
	}
	nMels := int64(r.Config.Features.NMels)
	out, err := r.enc.Run([]ort.Tensor{
		{F32: values, Shape: []int64{1, nMels, int64(frames)}},
		{I64: []int64{int64(r.feat.OutLen(len(wave)))}, Shape: []int64{1}},
	})
	if err != nil {
		return "", err
	}
	if len(out) < 2 || len(out[0].Shape) < 3 {
		return "", errors.New("энкодер вернул не то")
	}
	encD, encT := int(out[0].Shape[1]), int(out[0].Shape[2])
	encLen := encT
	if len(out[1].I64) > 0 {
		encLen = int(out[1].I64[0])
	} else if len(out[1].I32) > 0 {
		encLen = int(out[1].I32[0])
	}
	if encLen > encT {
		encLen = encT
	}
	ids, err := r.greedy(out[0].F32, encD, encT, encLen)
	if err != nil {
		return "", err
	}
	return r.tok.Decode(ids), nil
}

// Жадное декодирование: декодер пересчитывается только когда выдана буква.
func (r *Recognizer) greedy(encoded []float32, encD, encT, encLen int) ([]int, error) {
	blank := r.tok.BlankID()
	layers, hidden := r.Config.PredLayers, r.Config.PredHidden
	stateShape := []int64{int64(layers), 1, int64(hidden)}
	zeros := make([]float32, layers*hidden)
	var hyp []int
	label := blank
	h, c := zeros, zeros
	started := false
	var g, hNext, cNext []float32
	frame := make([]float32, encD)
	for t := 0; t < encLen; t++ {
		for d := 0; d < encD; d++ {
			frame[d] = encoded[d*encT+t]
		}
		for s := 0; s < maxSymbolsPerFrame; s++ {
			if g == nil {
				lab := blank
				hh, cc := zeros, zeros
				if started {
					lab, hh, cc = label, h, c
				}
				o, err := r.dec.Run([]ort.Tensor{
					{I64: []int64{int64(lab)}, Shape: []int64{1, 1}},
					{F32: hh, Shape: stateShape}, {F32: cc, Shape: stateShape},
				})
				if err != nil {
					return nil, err
				}
				g, hNext, cNext = o[0].F32, o[1].F32, o[2].F32
			}
			o, err := r.joint.Run([]ort.Tensor{
				{F32: frame, Shape: []int64{1, int64(encD), 1}},
				{F32: g, Shape: []int64{1, int64(hidden), 1}},
			})
			if err != nil {
				return nil, err
			}
			best, bestV := 0, float32(math.Inf(-1))
			for i, v := range o[0].F32 {
				if v > bestV {
					bestV, best = v, i
				}
			}
			if best == blank {
				break
			}
			hyp = append(hyp, best)
			label, h, c, started = best, hNext, cNext, true
			g = nil
		}
	}
	return hyp, nil
}
